File: src/main/java/net/alephdev/calendar/service/TaskService.java
package net.alephdev.calendar.service;

import jakarta.persistence.EntityNotFoundException;
import jakarta.transaction.Transactional;
import lombok.RequiredArgsConstructor;
import net.alephdev.calendar.dto.TaskDto;
import net.alephdev.calendar.models.*;
import net.alephdev.calendar.repository.RoleStatusRepository;
import net.alephdev.calendar.repository.StatusRepository;
import net.alephdev.calendar.repository.TaskRepository;
import net.alephdev.calendar.repository.functional.SprintRepository;

import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Sort;
import org.springframework.http.ResponseEntity;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.NoSuchElementException;
import java.util.Optional;

@Service
@RequiredArgsConstructor
public class TaskService {

    private final TaskRepository taskRepository;
    private final RoleStatusRepository roleStatusRepository;
    private final StatusRepository statusRepository;
    private final SprintRepository sprintRepository;
    private final UserService userService;
    private final RoleService roleService;


    public Page<Task> getAllTasks(int page) {
        return taskRepository.findAll(PageRequest.of(page, 20, Sort.by(Sort.Direction.DESC, "id")));
    }

    public Page<Task> getAllTaskByStatus(Integer statusId, int page) {
        return taskRepository.findAllByStatusId(statusId, PageRequest.of(page, 20, Sort.by(Sort.Direction.DESC, "id")));
    }

    public Page<Task> getAllTasksBySprint(Integer sprintId, int page) {
        return taskRepository.findAllBySprintId(sprintId, PageRequest.of(page, 20, Sort.by(Sort.Direction.DESC, "id")));
    }

    public Page<Task> getAllTasksByImplementer(User implementer, int page) {
        return taskRepository.findAllByImplementer(implementer, PageRequest.of(page, 20, Sort.by(Sort.Direction.DESC, "id")));
    }

    public Page<Task> getAllTasksBySprintAndStatus(Integer sprintId, Integer statusId, int page) {
        return taskRepository.findAllBySprintIdAndStatusId(sprintId, statusId, PageRequest.of(page, 20, Sort.by(Sort.Direction.DESC, "id")));
    }

    public Page<Task> getAllTasksByImplementerAndStatus(User implementer, Integer statusId, int page) {
        return taskRepository.findAllByImplementerAndStatusId(implementer, statusId, PageRequest.of(page, 20, Sort.by(Sort.Direction.DESC, "id")));
    }

    public Page<Task> getAllTasksByImplementerAndSprint(User implementer, Integer sprintId, int page) {
        return taskRepository.findAllByImplementerAndSprintId(implementer, sprintId, PageRequest.of(page, 20, Sort.by(Sort.Direction.DESC, "id")));
    }

    public Task createTask(TaskDto taskDto, User user) {
        Task task = new Task();
        task.setName(taskDto.getName().trim());
        task.setStoryPoints(taskDto.getStoryPoints());
        task.setPriorityEnum(taskDto.getPriorityEnum());
        task.setCreatedBy(user);
        task.setStatus(statusRepository.findById(1).orElseThrow(() -> new NoSuchElementException("Стандартный статус не найден")));

        return taskRepository.save(task);
    }

    public Task updateTask(Integer id, TaskDto updatedTask) {
        Task task = taskRepository.findById(id)
                .orElseThrow(() -> new NoSuchElementException("Задача не найдена"));

        if(updatedTask.getName() != null)  task.setName(updatedTask.getName().trim());
        if(updatedTask.getStoryPoints() != null) task.setStoryPoints(updatedTask.getStoryPoints());
        if(updatedTask.getPriorityEnum() != null) task.setPriorityEnum(updatedTask.getPriorityEnum());
        return taskRepository.save(task);
    }

    public Task getTask(Integer id) {
        return taskRepository.findById(id)
                .orElseThrow(() -> new NoSuchElementException("Задача не найдена"));
    }


    public ResponseEntity<Void> deleteTask(Integer id) {
        if (taskRepository.findById(id).isPresent()) {
            taskRepository.deleteById(id);
            return ResponseEntity.noContent().build();
        }
        return ResponseEntity.notFound().build();
    }

    public boolean canCreateTask(User user) {
        if (user.getRole() != null) {
            if(userService.isPrivileged(user)) {
                return true;
            }
            List<RoleStatus> roleStatuses = roleStatusRepository.findAllByRole_Id(user.getRole().getId());
            return roleStatuses.stream().anyMatch(rs -> rs.getStatus().getId() == 1);
        }
        return false;
    }


    public boolean canEditTask(User user, Task task) {
        if (userService.isPrivileged(user)) {
            return true;
        }
        return user.getLogin().equals(task.getCreatedBy().getLogin()) ||
                (task.getImplementer() != null && user.getLogin().equals(task.getImplementer().getLogin()));
    }

    @Transactional
    public Task assignImplementer(Integer taskId, String implementerLogin) {
        Task task = taskRepository.findById(taskId)
                .orElseThrow(() -> new NoSuchElementException("Задача не найдена"));
        try {
            User implementer = userService.getUserByLogin(implementerLogin);
            task.setImplementer(implementer);
        } catch (EntityNotFoundException e) {
            task.setImplementer(null);
        }

        return taskRepository.save(task);
    }

    @Transactional
    public Task updateStatus(Integer taskId, Integer statusId, User user) {
        Task task = taskRepository.findById(taskId)
                .orElseThrow(() -> new NoSuchElementException("Задача не найдена"));

        if (userService.isPrivileged(user)) {
            // Admin can set any status
            task.setStatus(statusRepository.findById(statusId).get());
        } else if (task.getImplementer() != null && task.getImplementer().getLogin().equals(user.getLogin()) && user.getRole() != null) {
            List<Status> allowedStatuses = roleService.getStatuses(user.getRole());
            Optional<Status> allowedStatus = allowedStatuses.stream()
                    .filter(roleStatus -> roleStatus.getId().equals(statusId))
                    .findFirst();

            if (allowedStatus.isPresent()) {
                task.setStatus(allowedStatus.get());
            } else {
                throw new IllegalArgumentException("Вы не можете ставить статус этой задаче");
            }
        } else {
            throw new IllegalArgumentException("Вы не можете обновлять статус этой задаче");
        }

        return taskRepository.save(task);
    }

    @Transactional
    public Task assignSprint(Integer taskId, Integer sprintId) {
        Task task = taskRepository.findById(taskId)
                .orElseThrow(() -> new NoSuchElementException("Задача не найдена"));

        try {
            Sprint sprint = sprintRepository.findById(sprintId)
                    .orElseThrow(() -> new NoSuchElementException("Спринт не найден"));
            task.setSprint(sprint);
        } catch (NoSuchElementException e) {
            task.setSprint(null);
        }

        return taskRepository.save(task);
    }
}

File: angular/src/app/services/server/task.service.ts
import {Injectable} from "@angular/core";
import {Observable} from "rxjs";
import {HttpClient, HttpErrorResponse} from "@angular/common/http";
import {ApiService} from "./api.service";
import {catchError} from "rxjs/operators";
import {Task} from "../../models/task";
import {TaskDto} from "../../models/dto/task-dto";
import {Page} from "../../models/misc/page";

@Injectable({
  providedIn: 'root'
})
export class TaskService {
  constructor(private http: HttpClient, private apiService: ApiService) {}

  getAllTasks(page: number = 0, statusId?: number, implementerLogin?: string, sprintId?: number): Observable<Page<Task> | HttpErrorResponse> {
    let params: { page: string, statusId?: string, implementerLogin?: string, sprintId?: string } = { page: page.toString() };
    if(statusId) params.statusId = statusId.toString();
    if(implementerLogin) params.implementerLogin = implementerLogin;
    if(sprintId) params.sprintId = sprintId.toString();

    return this.http.get<Page<Task>>(`${this.apiService.apiUrl}/tasks`, { params, headers: this.apiService.getHeaders() }).pipe(catchError(this.apiService.handleError));
  }


  createTask(taskDto: TaskDto): Observable<Task | HttpErrorResponse> {
    return this.http.post<Task>(`${this.apiService.apiUrl}/tasks`, taskDto, { headers: this.apiService.getHeaders() }).pipe(catchError(this.apiService.handleError));
  }

  updateTask(id: number, updatedTask: TaskDto): Observable<Task | HttpErrorResponse> {
    return this.http.put<Task>(`${this.apiService.apiUrl}/tasks/${id}`, updatedTask, { headers: this.apiService.getHeaders() }).pipe(catchError(this.apiService.handleError));
  }

  deleteTask(id: number): Observable<any | HttpErrorResponse> {
    return this.http.delete<any>(`${this.apiService.apiUrl}/tasks/${id}`, { headers: this.apiService.getHeaders() }).pipe(catchError(this.apiService.handleError));
  }


  assignImplementer(taskId: number, implementerLogin: string): Observable<Task | HttpErrorResponse> {
    return this.http.put<Task>(`${this.apiService.apiUrl}/tasks/${taskId}/implementer`, null, { params: { implementerLogin }, headers: this.apiService.getHeaders() }).pipe(catchError(this.apiService.handleError));
  }


  updateTaskStatus(taskId: number, statusId: number): Observable<Task | HttpErrorResponse> {
    return this.http.put<Task>(`${this.apiService.apiUrl}/tasks/${taskId}/status`, null, { params: { statusId: statusId.toString() }, headers: this.apiService.getHeaders() }).pipe(catchError(this.apiService.handleError));
  }


  assignSprintToTask(taskId: number, sprintId: number): Observable<Task | HttpErrorResponse> {
    return this.http.put<Task>(`${this.apiService.apiUrl}/tasks/${taskId}/sprint`, null, { params: { sprintId: sprintId.toString() }, headers: this.apiService.getHeaders() }).pipe(catchError(this.apiService.handleError));
  }
}
